teps/0118-matrix-with-explicit-combinations-of-parameters.md

```diff
@@ -0,0 +1,586 @@
+spec:
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - ...
+    - name: kaniko-build
+      taskRef:
+        name: kaniko
+      workspaces:
+        - name: source
+          workspace: shared-workspace
+      matrix:
+        - name: IMAGE
+          value: [ "image-1", "image-2", "image-3" ] 
+        - name: DOCKERFILE
+          value: [ "path/to/Dockerfile1", "path/to/Dockerfile2", "path/to/Dockerfile3" ]
+    - ...
+```
+
+`Matrix` by design with multiple `Parameters` calculates all the possible combinations and executes an instance of
+`PipelineTask` with each combination. Now, this specification of the `Matrix` creates 9 different combinations, such
+as:
+
+![Matrix for Kaniko with multiple Images and Dockerfile](images/0118-matrix-kaniko-with-images-and-dockerfile.jpg)
+
+This is not a desired outcome either. We need a way to specify explicit mapping between `IMAGE` and `DOCKERFILE`, such
+as:
+
+```yaml
+    - IMAGE: "image-1"
+      DOCKERFILE: "path/to/Dockerfile1"
+
+    - IMAGE: "image-2"
+      DOCKERFILE: "path/to/Dockerfile2"
+
+    - IMAGE: "image-3"
+      DOCKERFILE: "path/to/Dockerfile3"
+```
+
+### Goals
+
+The main goal of this TEP is to extend `Matrix` functionality to enable executing a `PipelineTask` with a limited or
+specific set of combinations of `Parameters`.
+
+### Non-Goals
+
+This TEP focuses on adding a specific combination of input values for `Matrix` parameters, excluding a specific
+combination from `Matrix` is out of scope for this TEP.
+
+## Proposal
+
+We propose adding a field - `include` - within the `matrix` field. This requires that we also move the existing array
+of `Parameters` to a field - `params` - within the `matrix` field. This allows users to specify the `Parameters`
+whose combinations should be generated, and the explicit combinations of `Parameters` that need to be added into the
+`Matrix`.
+
+### Include specific combinations in the Matrix
+
+Take the use case described [above](#adding-a-specific-combination-in-the-matrix). The `Pipeline` will need to define a
+`Matrix` that will execute `golang-test` in three different linux flavors. While running on `amd64` and `ppc64le`,
+`golang-test` should execute `go test github.com/tektoncd/pipeline/pkg/... -race -cover -v`. While running on
+`linux/s390x`, `golang-test` should execute `go test github.com/tektoncd/pipeline/pkg/... -cover -v`.
+
+The `Pipeline` addressing this use case will be defined as shown below:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline-to-build-and-test-go-project
+spec:
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - ...
+    - name: golang-test
+      taskRef:
+        name: golang-test
+      workspaces:
+        - name: source
+          workspace: shared-workspace
+      params:
+        - name: package
+          value: "github.com/tektoncd/pipeline"
+        - name: packages
+          value: "./pkg/..."
+      matrix:
+        params:
```

Comment:
`flags: ...` should be one of the matrix params here, right?

Reply:
nope, since `flags` is an options parameter with the default value of `-race -cover -v`. In this case, when the `flags` is not specified, the default value will be used to execute the `taskRun`.